[Sema] NFC: Refactor most TypeResolutionFlags into a traditional enum

davezarzycki opened the pull request:
`TypeResolutionFlags` is overly complicated at the moment because the vast majority of flag combinations are impossible and nonsensical. With this patch, we create a new `TypeResolverContext` type that is a classic enum and far easier to reason about. It also enables "exhaustive enum" checking, unlike the "flags" based approach.

davezarzycki committed:
[Sema] NFC: Refactor most TypeResolutionFlags into a traditional enum

TypeResolutionFlags is overly complicated at the moment because the vast
majority of flag combinations are impossible and nonsensical. With this
patch, we create a new TypeResolverContext type that is a classic enum
and far easier to reason about. It also enables "exhaustive enum"
checking, unlike the "flags" based approach.

davezarzycki commented:
@swift-ci please test

davezarzycki commented:
@swift-ci please smoke test linux

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -242,31 +240,31 @@ void TypeChecker::validateWhereClauses(ProtocolDecl *protocol,
-    options |= TypeResolutionFlags::GenericSignature;
-    options |= TypeResolutionFlags::InheritanceClause;
+    options = TypeResolutionOptions(TypeResolverContext::GenericSignature);
+    options.setContext(TypeResolverContext::InheritanceClause);
```
What does it mean to set the context twice like this?

davezarzycki replied:
Unless I'm missing something, `TypeResolutionOptions` is about providing context to nested type resolvers. Some type resolvers need to know the current context, some need to know the outermost "base" context, and some need to know both. After this pull request, the `TypeResolutionOptions` constructor sets both the base and current contexts to the provided `TypeResolverContext`. When `setContext()` is called, only the current context is updated. 

In the case of this specific code, `GenericSignature` doesn't make sense as the sub-context of another `TypeResolverContext` and therefore it is setup as the "base" context in this example. That being said, we need to convey that the inheritance clause needs to be checked too, and therefore it is set as the "current" context so that it can be special cased accordingly.

slavapestov replied:
I see, thanks. It looks like InheritanceClause is only checked in one place — and there’s a FIXME about it. Might be worth trying to remove it altogether.

davezarzycki replied:
Weird. I commented out the exception and nothing broke on my Linux box. Is that expected?

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -242,31 +240,31 @@ void TypeChecker::validateWhereClauses(ProtocolDecl *protocol,
     if (auto nominal = dyn_cast<NominalTypeDecl>(DC)) {
       DC = nominal;
-      options |= TypeResolutionFlags::GenericSignature;
+      options = TypeResolutionOptions(TypeResolverContext::GenericSignature);
```
An implicit conversion from TypeResolverContext to TypeResolutionOptions might be nice

davezarzycki replied:
Sure, that can work. FWIW – I was avoiding implicit conversions in the short term to make this refactoring work easier to reason about.

davezarzycki replied:
After some thought, I'd prefer that we do not allow implicit conversion from `TypeResolverContext` to `TypeResolutionOptions` during assignment. Construction is okay though. The problem is that implicit conversion during assignment invites bugs, because everything is being reset (the base, the context, and flags).

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -244,15 +244,15 @@ void TypeChecker::checkGenericParamList(GenericSignatureBuilder *builder,
   // Determine where and how to perform name lookup for the generic
   // parameter lists and where clause.
-  TypeResolutionOptions options;
+  TypeResolutionOptions options = None;
```
Instead of None, why not have a 'default' context?

davezarzycki replied:
I wanted to avoid error-of-omission bugs.

slavapestov commented on lib/Sema/TypeCheckPattern.cpp:
```diff
@@ -1024,7 +1025,7 @@ bool TypeChecker::coercePatternToType(Pattern *&P, DeclContext *dc, Type type,
     // labeled tuple, treat this as a tuple pattern. It's unlikely that the
     // user is interested in binding a variable of type (foo: Int).
-    if ((options & TypeResolutionFlags::EnumPatternPayload)
+    if ((options.getContext() == TypeResolverContext::EnumPatternPayload)
```
maybe options.is(TypeResolverContext::EnumPatternPayload) would read better?

davezarzycki replied:
Sure. Just FYI – that won't remove the need for `getContext()` thought (for use with `switch` statements, etc).

slavapestov commented on lib/Sema/TypeCheckRequestFunctions.cpp:
```diff
@@ -25,13 +25,13 @@ Type InheritedTypeRequest::evaluate(
-      options |= TypeResolutionFlags::GenericSignature;
-      options |= TypeResolutionFlags::InheritanceClause;
+      options = TypeResolutionOptions(TypeResolverContext::GenericSignature);
+      options.setContext(TypeResolverContext::InheritanceClause);
```
Once again you're setting the context twice?

davezarzycki replied:
See the earlier explanation.

slavapestov commented:
Overall I really like the idea of making the flags more semantic. I think subsequent refactoring could be done to pare down the number of flags too. In the very long term, we might be able to derive the “context” entirely from source locations even.

davezarzycki commented:
Ya, I like this change too. The “everything is a flag” model is brittle and has a steep learning curve.

davezarzycki committed:
[Sema] NFC: Remove dead code: TypeResolverContext::InheritanceClause

davezarzycki commented:
I think I addressed all of the feedback. I also renamed some of the `TypeResolverContext` enum cases to be aligned with the AST node name to be avoid needless renaming. Finally, a second commit attempts to address the FIXME that @slavapestov pointed out.

@swift-ci Please test source compatibility
@swift-ci please test macos
@swift-ci please smoke test linux

davezarzycki commented:
@swift-ci Please test source compatibility

davezarzycki commented:
@swift-ci please test

slavapestov commented:
Sorry, do you mind resolving the merge conflicts? Then the change LGTM

davezarzycki commented:
@swift-ci please smoke test and merge

davezarzycki commented:
@swift-ci please smoke test and merge

DougGregor commented:
Thank you, this is much cleaner.
